wave_vision/include/wave/vision/tracker/impl/tracker.hpp

```diff
@@ -0,0 +1,239 @@
+              curr_time, sensor_id, curr_ids.at(m.trainIdx), curr_landmark);
+
+            // Add to container
+            this->landmarks.insert(prev_measurement);
+            this->landmarks.insert(curr_measurement);
+        }
+    }
+
+    return curr_ids;
+}
+
+// Public Functions
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+std::vector<FeatureTrack> Tracker<TDetector, TDescriptor, TMatcher>::getTracks(
+  const size_t &img_num) const {
+    std::vector<FeatureTrack> feature_tracks;
+    std::vector<size_t> landmark_ids;
+
+    // TODO add in ability to choose different camera.
+    int sensor_id = 0;
+
+    if (img_num > this->img_count) {
+        throw std::out_of_range("Image requested is in the future!");
+    } else if (img_num > 0) {
+        // Find the time for this image
+        std::chrono::steady_clock::time_point img_time =
+          this->img_times.at(img_num);
+
+        // Extract all of the IDs visible at this time
+        landmark_ids =
+          this->landmarks.getLandmarkIDsInWindow(img_time, img_time);
+
+        // For each ID, get the track.
+        for (const auto &l : landmark_ids) {
+            // Looking for track from first image.
+            std::chrono::steady_clock::time_point start_time =
+              (this->img_times.begin())->second;
+
+            std::vector<LandmarkMeasurement<int>> tracks =
+              this->landmarks.getTrackInWindow(
+                sensor_id, l, start_time, img_time);
+
+            // Find image corresponding to start time
+            size_t first_image;
+
+            // Last image feature was seen in must be image requested.
+            size_t last_image = img_num;
+
+            // Iterators for time map and landmark measurements
+            std::map<size_t,
+                     std::chrono::steady_clock::time_point>::const_iterator
+              times;
+            std::vector<LandmarkMeasurement<int>>::const_iterator first =
+              tracks.begin();
+
+            // Check if the time in the map corresponds to the time of the first
+            // landmark measurement
+            for (times = this->img_times.begin();
+                 times != this->img_times.end();
+                 ++times) {
+                if (times->second == first->time_point) {
+                    first_image = times->first;
+                }
+            }
+
+            // Emplace new feature track back into vector
+            feature_tracks.emplace_back(l, tracks, first_image, last_image);
+        }
+    } else {
+        // img_num is zero, and there are no tracks.
+        feature_tracks = {};
+    }
+
+    return feature_tracks;
+}
+
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+void Tracker<TDetector, TDescriptor, TMatcher>::addImage(const cv::Mat &image) {
+    // Get time for keypoint registration
+    std::chrono::steady_clock::time_point current_time = clock.now();
```

Comment:
I'm not sure if this makes sense to get the time here as well. Actually, I just realized this should read `this->clock.now();`.

But does that even make sense in an online setting? This is currently set up as a member clock. I think whether it's online or offline, it should take a time from elsewhere.

But for offline, times maybe as a whole don't make sense, which you've mentioned below.